File: src-tauri/src/services/notification_service.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};

const MAX_NOTIFICATIONS: usize = 500;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Notification {
    pub id: u64,
    #[serde(rename = "type")]
    pub notification_type: String,
    pub title: String,
    pub body: String,
    pub session_id: Option<String>,
    pub timestamp: String,
    pub read: bool,
}

#[derive(Debug, Serialize, Deserialize)]
struct NotificationState {
    notifications: Vec<Notification>,
    #[serde(rename = "nextId")]
    next_id: u64,
}

/// Incoming notification from a JSON file dropped by Copilot CLI.
#[derive(Debug, Deserialize)]
struct IncomingNotification {
    #[serde(rename = "type", default)]
    notification_type: Option<String>,
    #[serde(default)]
    title: Option<String>,
    #[serde(default)]
    body: Option<String>,
    #[serde(rename = "sessionId", default)]
    session_id: Option<String>,
    #[serde(default)]
    timestamp: Option<String>,
}

pub struct NotificationService {
    notifications_dir: PathBuf,
    state_file: PathBuf,
    notifications: Vec<Notification>,
    next_id: u64,
    processed_files: std::collections::HashSet<String>,
}

impl NotificationService {
    pub fn new(notifications_dir: &Path) -> Self {
        let _ = fs::create_dir_all(notifications_dir);
        let state_file = notifications_dir.join(".state.json");

        let mut svc = Self {
            notifications_dir: notifications_dir.to_path_buf(),
            state_file,
            notifications: Vec::new(),
            next_id: 1,
            processed_files: std::collections::HashSet::new(),
        };
        svc.load_state();
        svc.scan_existing();
        svc
    }

    fn load_state(&mut self) {
        if let Ok(data) = fs::read_to_string(&self.state_file) {
            if let Ok(state) = serde_json::from_str::<NotificationState>(&data) {
                self.notifications = state.notifications;
                self.next_id = state.next_id;
            }
        }
    }

    fn save_state(&self) {
        let mut notifications = self.notifications.clone();
        if notifications.len() > MAX_NOTIFICATIONS {
            notifications = notifications.split_off(notifications.len() - MAX_NOTIFICATIONS);
        }

        let state = NotificationState {
            notifications,
            next_id: self.next_id,
        };

        if let Ok(json) = serde_json::to_string_pretty(&state) {
            let _ = fs::write(&self.state_file, json);
        }
    }

    fn scan_existing(&mut self) {
        if let Ok(entries) = fs::read_dir(&self.notifications_dir) {
            let files: Vec<_> = entries
                .flatten()
                .filter(|e| {
                    let name = e.file_name().to_string_lossy().to_string();
                    name.ends_with(".json") && name != ".state.json"
                })
                .collect();

            for entry in files {
                let filename = entry.file_name().to_string_lossy().to_string();
                self.process_file(&filename);
            }
        }
    }

    fn process_file(&mut self, filename: &str) {
        if self.processed_files.contains(filename) {
            return;
        }

        let file_path = self.notifications_dir.join(filename);
        if !file_path.exists() {
            return;
        }

        match fs::read_to_string(&file_path) {
            Ok(content) => {
                if let Ok(data) = serde_json::from_str::<IncomingNotification>(&content) {
                    let notification = Notification {
                        id: self.next_id,
                        notification_type: data.notification_type.unwrap_or_else(|| "info".to_string()),
                        title: data.title.unwrap_or_else(|| "Notification".to_string()),
                        body: data.body.unwrap_or_default(),
                        session_id: data.session_id,
                        timestamp: data
                            .timestamp
                            .unwrap_or_else(|| chrono::Utc::now().to_rfc3339()),
                        read: false,
                    };
                    self.next_id += 1;
                    self.notifications.push(notification);
                    self.save_state();
                }
            }
            Err(_) => {}
        }

        self.processed_files.insert(filename.to_string());

        // Cap processed files set
        if self.processed_files.len() > 1000 {
            let excess: Vec<_> = self.processed_files.iter().take(500).cloned().collect();
            for f in excess {
                self.processed_files.remove(&f);
            }
        }

        // Consume the file
        let _ = fs::remove_file(&file_path);
    }

    /// Push a notification programmatically (e.g., on session exit).
    pub fn push(&mut self, notification_type: &str, title: &str, body: &str, session_id: Option<&str>) -> Notification {
        let n = Notification {
            id: self.next_id,
            notification_type: notification_type.to_string(),
            title: title.to_string(),
            body: body.to_string(),
            session_id: session_id.map(|s| s.to_string()),
            timestamp: chrono::Utc::now().to_rfc3339(),
            read: false,
        };
        self.next_id += 1;
        self.notifications.push(n.clone());
        self.save_state();
        n
    }

    pub fn get_all(&self) -> Vec<Notification> {
        self.notifications.clone()
    }

    pub fn get_unread_count(&self) -> u64 {
        self.notifications.iter().filter(|n| !n.read).count() as u64
    }

    pub fn mark_read(&mut self, id: u64) {
        if let Some(n) = self.notifications.iter_mut().find(|n| n.id == id) {
            n.read = true;
            self.save_state();
        }
    }

    pub fn mark_all_read(&mut self) {
        for n in &mut self.notifications {
            n.read = true;
        }
        self.save_state();
    }

    pub fn dismiss(&mut self, id: u64) {
        self.notifications.retain(|n| n.id != id);
        self.save_state();
    }

    pub fn clear_all(&mut self) {
        self.notifications.clear();
        self.save_state();
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::fs;

    fn temp_dir() -> tempfile::TempDir {
        tempfile::tempdir().unwrap()
    }

    #[test]
    fn test_push_and_get() {
        let dir = temp_dir();
        let mut svc = NotificationService::new(dir.path());
        svc.push("task-done", "Build passed", "exit 0", Some("sess-1"));

        let all = svc.get_all();
        assert_eq!(all.len(), 1);
        assert_eq!(all[0].title, "Build passed");
        assert_eq!(all[0].notification_type, "task-done");
        assert_eq!(all[0].session_id, Some("sess-1".to_string()));
        assert!(!all[0].read);
    }

    #[test]
    fn test_unread_count() {
        let dir = temp_dir();
        let mut svc = NotificationService::new(dir.path());
        svc.push("info", "One", "", None);
        svc.push("info", "Two", "", None);
        assert_eq!(svc.get_unread_count(), 2);

        svc.mark_read(1);
        assert_eq!(svc.get_unread_count(), 1);
    }

    #[test]
    fn test_mark_all_read() {
        let dir = temp_dir();
        let mut svc = NotificationService::new(dir.path());
        svc.push("info", "One", "", None);
        svc.push("info", "Two", "", None);
        svc.mark_all_read();
        assert_eq!(svc.get_unread_count(), 0);
    }

    #[test]
    fn test_dismiss() {
        let dir = temp_dir();
        let mut svc = NotificationService::new(dir.path());
        svc.push("info", "One", "", None);
        svc.push("info", "Two", "", None);
        svc.dismiss(1);
        assert_eq!(svc.get_all().len(), 1);
        assert_eq!(svc.get_all()[0].title, "Two");
    }

    #[test]
    fn test_clear_all() {
        let dir = temp_dir();
        let mut svc = NotificationService::new(dir.path());
        svc.push("info", "One", "", None);
        svc.push("info", "Two", "", None);
        svc.clear_all();
        assert!(svc.get_all().is_empty());
    }

    #[test]
    fn test_state_persistence() {
        let dir = temp_dir();
        {
            let mut svc = NotificationService::new(dir.path());
            svc.push("error", "Session crashed", "exit 1", Some("s1"));
            svc.mark_read(1);
        }
        {
            let svc = NotificationService::new(dir.path());
            let all = svc.get_all();
            assert_eq!(all.len(), 1);
            assert_eq!(all[0].title, "Session crashed");
            assert!(all[0].read);
        }
    }

    #[test]
    fn test_consumes_json_files() {
        let dir = temp_dir();
        let notif_json = serde_json::json!({
            "type": "task-done",
            "title": "Build complete",
            "body": "All tests passed",
            "sessionId": "abc-123"
        });
        fs::write(dir.path().join("notif-001.json"), notif_json.to_string()).unwrap();

        let svc = NotificationService::new(dir.path());
        let all = svc.get_all();
        assert_eq!(all.len(), 1);
        assert_eq!(all[0].title, "Build complete");

        // File should be consumed (deleted)
        assert!(!dir.path().join("notif-001.json").exists());
    }
}

File: src/notification-service.js
const fs = require('fs');
const path = require('path');
const EventEmitter = require('events');

const MAX_NOTIFICATIONS = 500;
const MAX_PROCESSED_FILES = 1000;

class NotificationService extends EventEmitter {
  constructor(notificationsDir) {
    super();
    this.notificationsDir = notificationsDir;
    this.watcher = null;
    this.processedFiles = new Set();
    this.notifications = []; // { id, type, title, body, sessionId, timestamp, read }
    this.nextId = 1;
    this.stateFile = path.join(notificationsDir, '.state.json');

    // Ensure directory exists
    fs.mkdirSync(notificationsDir, { recursive: true });
    this._loadState();
  }

  start() {
    // Process any existing notification files on startup
    this._scanExisting();

    // Watch for new files
    try {
      this.watcher = fs.watch(this.notificationsDir, (eventType, filename) => {
        if (!filename || !filename.endsWith('.json') || filename === '.state.json') return;
        if (eventType === 'rename' || eventType === 'change') {
          // Small delay to ensure file is fully written
          setTimeout(() => this._processFile(filename), 100);
        }
      });
      this.watcher.on('error', (err) => {
        console.error('[notifications] Watcher error:', err.message);
        this.watcher = null;
      });
    } catch (err) {
      console.error('[notifications] Failed to start watcher:', err.message);
    }
  }

  stop() {
    if (this.watcher) {
      this.watcher.close();
      this.watcher = null;
    }
    this._saveState();
  }

  _scanExisting() {
    try {
      const files = fs.readdirSync(this.notificationsDir)
        .filter(f => f.endsWith('.json') && f !== '.state.json');
      for (const file of files) {
        this._processFile(file);
      }
    } catch {}
  }

  _processFile(filename) {
    if (this.processedFiles.has(filename)) return;
    const filePath = path.join(this.notificationsDir, filename);
    try {
      if (!fs.existsSync(filePath)) return;
      const content = fs.readFileSync(filePath, 'utf8');
      const data = JSON.parse(content);

      const notification = {
        id: this.nextId++,
        type: data.type || 'info',
        title: data.title || 'Notification',
        body: data.body || '',
        sessionId: data.sessionId || null,
        timestamp: data.timestamp || new Date().toISOString(),
        read: false,
      };

      this.notifications.push(notification);
      this.emit('notification', notification);
      this._saveState();

      this.processedFiles.add(filename);

      // Cap processed files set to prevent unbounded growth
      if (this.processedFiles.size > MAX_PROCESSED_FILES) {
        const excess = this.processedFiles.size - MAX_PROCESSED_FILES;
        const iter = this.processedFiles.values();
        for (let i = 0; i < excess; i++) {
          this.processedFiles.delete(iter.next().value);
        }
      }

      // Consume the file
      try { fs.unlinkSync(filePath); } catch {}
    } catch (err) {
      // Invalid JSON or read error — delete the bad file
      try { fs.unlinkSync(filePath); } catch {}
    }
  }

  // Push a notification programmatically (from GUI state detection)
  push(notification) {
    const n = {
      id: this.nextId++,
      type: notification.type || 'info',
      title: notification.title || 'Notification',
      body: notification.body || '',
      sessionId: notification.sessionId || null,
      timestamp: notification.timestamp || new Date().toISOString(),
      read: false,
    };
    this.notifications.push(n);
    this.emit('notification', n);
    this._saveState();
    return n;
  }

  markRead(id) {
    const n = this.notifications.find(n => n.id === id);
    if (n) { n.read = true; this._saveState(); }
  }

  markAllRead() {
    this.notifications.forEach(n => { n.read = true; });
    this._saveState();
  }

  dismiss(id) {
    this.notifications = this.notifications.filter(n => n.id !== id);
    this._saveState();
  }

  clearAll() {
    this.notifications = [];
    this._saveState();
  }

  getAll() {
    return [...this.notifications];
  }

  getUnreadCount() {
    return this.notifications.filter(n => !n.read).length;
  }

  getUnreadCountForSession(sessionId) {
    return this.notifications.filter(n => !n.read && n.sessionId === sessionId).length;
  }

  _loadState() {
    try {
      if (fs.existsSync(this.stateFile)) {
        const data = JSON.parse(fs.readFileSync(this.stateFile, 'utf8'));
        this.notifications = Array.isArray(data.notifications) ? data.notifications : [];
        this.nextId = typeof data.nextId === 'number' ? data.nextId : 1;
      }
    } catch {
      this.notifications = [];
      this.nextId = 1;
    }
  }

  _saveState() {
    try {
      if (this.notifications.length > MAX_NOTIFICATIONS) {
        this.notifications = this.notifications.slice(-MAX_NOTIFICATIONS);
      }
      fs.writeFileSync(this.stateFile, JSON.stringify({
        notifications: this.notifications,
        nextId: this.nextId,
      }, null, 2), 'utf8');
    } catch {}
  }
}

module.exports = NotificationService;
